### ISSUE TITLE
Diagram showing network concepts mapped to a data center

### ISSUE THREAD

**Issue by degville** [CanonicalLtd/maas-docs#581]
Taken from:
https://github.com/CanonicalLtd/maas-docs/issues/573

This is feedback from the field as people find "fabrics vs spaces" confusing given spaces became an L2 concept in 2.2.0:
https://github.com/CanonicalLtd/maas-docs/blame/2.2/en/intro-concepts.md#L127-L131

With network spaces there are several things in-place:

1:M VLANs per space
1:N subnets per VLAN
1:K fabrics per space (in other words: VLANs can be on different spaces)
But in the end it's a group of VLANs with a symbolic name where routing between the respective subnets works (e.g. via router-on-a-stick mechanism).

It would be nice to have a diagram as well showing all of those concepts mapped to a data center.

**Comment by dshcherb** [comment]
I'm going to make a draft diagram (or several diagrams) to illustrate those concepts in different variations.

**Comment by dshcherb** [comment]
Two diagrams illustrating fabrics.

I'll add more for spaces.
- https://github.com/dshcherb/maas-concepts/blob/master/fabrics.svg
- https://github.com/dshcherb/maas-concepts/blob/master/fabrics-abstract.svg

**Comment by dshcherb** [comment]
- VLANs belong to fabrics
- a space is a group of VLANs
- subnets belong to VLANs AND to spaces (transitively)

So, you may have a space that consists of multiple VLANs in different fabrics (VLANs may even have the **same number** in different fabrics) and different subnets belonging to that space.

Having the same subnet in different VLANs does not make much sense within a single routing domain (https://en.wikipedia.org/wiki/Routing_domain, https://www.iso.org/obp/ui#iso_std_iso-iec_tr_9575_ed-2_v1_en_term_3.4.2) because it just complicates the way you build your L3 network. Normally, different VLANs will have different subnets and that is why network spaces allow for multiple VLANs with multiple different subnets.

Different VLANs may have same subnets provided that NAT is implemented at certain points in the L3 network which avoids certain complications in routing. Otherwise one has to deal with overlapping subnets, multiple routing tables per host and policy-based routing.

Those cases are not unique because there is a limited number of private IPv4 address ranges and sometimes companies merge and have to somehow resolve these issues.

- https://tools.ietf.org/html/rfc1918#section-3
- https://www.cisco.com/c/en/us/support/docs/ip/network-address-translation-nat/13774-3.html

In the IaaS world overlay networks do exactly that: they provide virtual fabrics (~ L2 networks, ~ broadcast domains) terminated by a NAT (usually static NAT or port address translation).

Our docs make many hidden assumptions without references to RFCs and best practices which is why it is complicated to understand why additional terms are introduced and how they map to other common networking terms.

**Comment by degville** [comment]
Thank you so much for taking the trouble do create these diagrams, and providing the extra fabric details. They'll help make understanding these concepts much easier! We'll likely tweak them slightly and roll them into the docs asap.

**Comment by dshcherb** [comment]
This diagram is for spaces.

https://github.com/dshcherb/maas-concepts/blob/master/spaces.svg

> This diagram also includes spaces which are logical groupings of VLANs with subnets that are able to "communicate with each other". "communicate" is a bit of a vague description. In practice it means that there are layer 3 devices that are able to forward packets from each VLAN to all other VLANs (an L3 switch or a router), moreover, these L3 devices may or may not be directly connected to those VLANs. Hosts may either use that L3 device as a default gateway or have static routes, however, that next hop L3 device must be able to send L3 packets to the right destination as there is no other way to go from one L2 (broadcast domain) to the other.
> Notice that fabrics and physical VLAN placement become irrelevant here - the fact that we use VLAN 10 on different fabrics does not concern us. Subnets are different in every VLAN. Some could be the same but this is a very complex network scenario which includes overlapping subnets and careful static route management and policy-based routing.
> Routes are summarized via green lines, however, each layer 3 device must have a next hop to the final subnet - this is not just about end devices.

Additional notes about the diagram and the "able to communicate" definition which needs a proper explanation:

https://en.wikipedia.org/wiki/Hop_(networking)
"hop is one portion of the path between source and destination"
"Each router along the data path constitutes a hop, as the data is moved from one Layer 3 network to another. Hop count is therefore a basic measurement of distance in a network."

Now to this definition:
"A space is a logical grouping of VLANs whose subnets are able to communicate with one another."

In general:

- one [broadcast domain](https://en.wikipedia.org/wiki/Broadcast_domain) ~ one VLAN : N subnets, or, in other words, one VLAN can have many subnets (this case is not modeled by MAAS - you can only assign a single subnet per VLAN in MAAS) because frames can carry any L3 traffic regardless of L3 network addresses.
- subnet_1 <-> hop_1 <-> ... <-> hop_n <-> subnet_2, or: there can be many hops along the layer 3 network path.
- A path from a host in subnet_1 to a host in subnet_2 may not be the same as a path from the host in subnet_2 to the host in subnet_1.


So, to summarize:

A space is a logical grouping of VLANs where packets sent by hosts in their respective subnets are able to reach any other host in any other subnet via layer 3 path(s) which

I think referring to the [OSI layers](https://en.wikipedia.org/wiki/OSI_model) may be normal for anybody who has basic experience in computer networking.

Referring to graphs, having a multi-hop hop path is also called network reachability. 

https://tools.ietf.org/html/rfc5549
"IPv4 Network Layer Reachability"

**Comment by rezroo** [comment]
I'm inheriting 10 servers with 2 10G access switches and a 1G core switch. The network given to access these from the corporate is a 172.xxx.xxx.1/25 CIDR. The current configuration of the switches and wiring has some of the nodes (control nodes) connected directly to the 1G for access, and the others are connected on a vlan through 10G port - no 1G access for /25 network. So I tried to define the /25 subnet twice, one on the 1G and a second time on the 10G on a vlan - and MaaS says that the subnet is already defined - even though I'm defining on a different space and fabric. Is there any way to tell MaaS to do the what I want?

**Comment by dshcherb** [comment]
@rezroo

I would expect that to work. The best place to file a bug is launchpad as devs look there and this is merely a doc repo.

https://bugs.launchpad.net/maas/+filebug

Before doing that check which version do you have and if there are any logs you can provide.

Outputs of commands below would be helpful.

maas <profile> fabrics read
maas <profile> machines read
maas <profile> subnets read
maas <profile> spaces read

maas <profile> version read
dpkg -l '*maas*'

Do you really need to use /25 on both VLANs though? Maybe it's worthwhile to partition it into smaller segments and assign a subnet per vlan. to avoid having to deal with overlapping subnets and syncing reservations on both?

https://bugs.launchpad.net/maas/+bug/1683047
